OpenRA.Mods.Common/Traits/BotModules/McvExpansionManagerBotModule.cs

```diff
@@ -0,0 +1,932 @@
+					.ToList();
+
+				if (conyards.Count > 1)
+				{
+					var movableMCV = conyards.FirstOrDefault(a => !a.TraitsImplementing<ProductionQueue>()
+					.Any(t => t.Enabled && t.AllQueued().Any(q => Info.RefineryTypes.Contains(q.Item))));
+
+					if (movableMCV != null)
+						bot.QueueOrder(new Order("DeployTransform", movableMCV, true));
+				}
+			}
+		}
+
+		// Find any MCV and deploy them at a sensible location.
+		void DeployMcv(IBot bot, Actor mcv, bool move)
+		{
+			if (move)
+			{
+				var transformsInfo = mcv.Info.TraitInfo<TransformsInfo>();
+				var desiredLocation = ChooseMcvDeployLocation(mcv, transformsInfo.IntoActor, transformsInfo.Offset, allowfallback);
+				if (desiredLocation == null)
+					return;
+
+				bot.QueueOrder(new Order("Move", mcv, Target.FromCell(world, desiredLocation.Value), true));
+
+				allowfallback = true;
+
+				if (constructionYards.Actors.Any(a => !a.IsDead))
+				{
+					foreach (var n in notifyPositionsUpdated)
+					{
+						n.UpdatedBaseCenter(desiredLocation.Value);
+						n.UpdatedDefenseCenter(desiredLocation.Value);
+					}
+				}
+			}
+
+			bot.QueueOrder(new Order("DeployTransform", mcv, true));
+		}
+
+		// First, find a suitable expansion location according to current mode,
+		// Then, find a deployable cell around it.
+		CPos? ChooseMcvDeployLocation(Actor mcv, string transformIntoType, CVec offset, bool allowfallback)
+		{
+			var actorInfo = world.Map.Rules.Actors[transformIntoType];
+			var bi = actorInfo.TraitInfoOrDefault<BuildingInfo>();
+			if (bi == null)
+				return null;
+
+			Mobile mobile = null;
+			if (mcv.TraitOrDefault<IMove>() == null)
+				return null;
+			else
+				mobile = mcv.TraitOrDefault<Mobile>();
+
+			var (expandCenter, attraction, checkspot, pfcount) = GetExpansionCenter(mcv, mobile, allowfallback);
+
+			// Find the deployable cell
+			CPos? FindDeployCell(CPos? sourceCell, CPos? targetCell, int minRange, int maxRange, int tryMaintainRange, int pfcount)
+			{
+				if (!sourceCell.HasValue || !targetCell.HasValue)
+					return null;
+
+				var target = targetCell.Value;
+				var source = sourceCell.Value;
+
+				var cells = world.Map.FindTilesInAnnulus(target, minRange, maxRange).Where(c => world.CanPlaceBuilding(c + offset, actorInfo, bi, null));
+
+				/* First, sort the cells that keep tryMaintainRange to target (meanwhile direction is from center to target) the first to be considered
+				 * by using following code. The idea is to use a linear combination of two distances-square for sorting weight.
+				 *
+				 * See comments in https://github.com/OpenRA/OpenRA/pull/22028#issuecomment-3242518793 for explaination.
+				 */
+				if (source != target)
+				{
+					var theta = tryMaintainRange;
+					var deta = (pfcount < 0 ? (source - target).Length : pfcount) - tryMaintainRange;
+
+					return cells.OrderBy(c =>
+					{
+						var c2target = pathfinder.FindPathToTargetCells(mcv, c, [target], BlockedByActor.Immovable);
```

Comment:
@RoosterDragon Actually this is the function I worry the most, after I get an expansion cell like a resource piece, I have to find a place near it in pf.

Reply:
but if not, we may encounter strange state like this https://github.com/OpenRA/OpenRA/pull/22028#pullrequestreview-3196670334